src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php

```diff
@@ -1376,6 +1376,9 @@ private function getServiceCall($id, Reference $reference = null)
             return '$this';
         }
 
+        if ($this->container->hasDefinition($id) && !$this->container->getDefinition($id)->isPublic()) {
+            return "\$this->services[(isset(\$this->services['$id']) || \$this->{$this->generateMethodName($id)}()) && false ?: '$id']";
```

Comment:
this code looks very hard to understand ? Why not keeping your previous proposal but simply removing the assignment ? the factory method handles the assignment (to share the service) but also returns it.